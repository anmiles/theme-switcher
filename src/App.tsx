import { useState } from 'react';

import { UserProvider } from './lib/providers/userProvider';
import { SystemProvider } from './lib/providers/systemProvider';
import './styles/themeSwitcher.css';

interface AppProps {
	readonly svgRoot : string;
}

export default function App({ svgRoot }: AppProps) {
	const userProvider   = new UserProvider();
	const systemProvider = new SystemProvider();

	const [ userTheme, setUserTheme ]     = useState(userProvider.get());
	const [ systemTheme, setSystemTheme ] = useState(systemProvider.get());

	const theme = userTheme ?? systemTheme;
	document.body.setAttribute('data-theme', theme);

	userProvider.on('change', setUserTheme);
	systemProvider.on('change', setSystemTheme);
	systemProvider.watch();

	return (
		<div
			className="themeSwitcher"
			onClick={ () => {
				userProvider.changeTheme(userTheme);
			} }
		>
			{ userTheme === undefined && <svg><use href={ `${svgRoot}/system.svg#svg` } /></svg> }
			{ userTheme === 'light' && <svg><use href={ `${svgRoot}/light.svg#svg` } /></svg> }
			{ userTheme === 'dark' && <svg><use href={ `${svgRoot}/dark.svg#svg` } /></svg> }
		</div>
	);
}
