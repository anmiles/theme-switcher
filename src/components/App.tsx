import type { CSSProperties } from 'react';
import { useEffect, useMemo, useState } from 'react';

import { UserProvider } from '../providers/userProvider';
import { SystemProvider } from '../providers/systemProvider';
import '../styles/style.css';
import Icon from './Icon';
import ThemeSelector from './ThemeSelector';

export interface AppProps {
	readonly float? : CSSProperties['float'];
}

export default function App({ float }: AppProps) {
	const userProvider   = useMemo(() => new UserProvider(), []);
	const systemProvider = useMemo(() => new SystemProvider(), []);

	const [ userTheme, setUserTheme ]     = useState(userProvider.get());
	const [ systemTheme, setSystemTheme ] = useState(systemProvider.get());
	const [ showList, setShowList ]       = useState(false);

	const theme = userTheme ?? systemTheme;

	useEffect(() => {
		document.body.setAttribute('data-theme', theme);
		userProvider.on('change', setUserTheme);
		systemProvider.on('change', setSystemTheme);
		systemProvider.watch();
	}, [ theme, userProvider, systemProvider ]);

	return (
		<div
			className="themeSwitcher"
			data-testid="theme-switcher"
			data-float={ float }
			onClick={ () => {
				setShowList(!showList);
			} }
		>
			<Icon theme={ theme } />

			{ !showList
				? null
				: (
					<ThemeSelector
						currentUserTheme={ userTheme }
						onListItemClick={ (theme) => {
							userProvider.set(theme);
							setShowList(false);
						} }
					/>
				) }
		</div>
	);
}
