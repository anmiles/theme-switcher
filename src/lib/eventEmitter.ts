class EventEmitter<TEventMap extends Record<string, Array<unknown>>> {
	private readonly listeners : {
		[TEvent in keyof TEventMap]?: Array<(...data: TEventMap[TEvent]) => void>
	} = {};

	public on<TEvent extends keyof TEventMap>(
		event: TEvent,
		listener: (...data: TEventMap[TEvent]) => void,
	): void {
		const listeners = this.listeners[event] ??= [];
		listeners.push(listener);
	}

	public off<TEvent extends keyof TEventMap>(
		event: TEvent,
		listener: (...data: TEventMap[TEvent]) => void,
	): void {
		const listeners = this.listeners[event] ??= [];
		listeners.splice(listeners.indexOf(listener), 1);
	}

	protected emit<TEvent extends keyof TEventMap>(
		event: TEvent,
		...data: TEventMap[TEvent]
	): void {
		this.listeners[event]?.forEach((listener) => {
			listener(...data);
		});
	}
}

export { EventEmitter };
