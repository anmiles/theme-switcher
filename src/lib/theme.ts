const themes = [ 'light', 'dark' ] as const;
type Theme = typeof themes[number];
const defaultTheme: Theme = 'light';

function isTheme(arg: unknown): arg is Theme {
	return typeof arg === 'string' && themes.includes(arg as Theme);
}

function getThemeName(theme: Theme | undefined): string {
	switch (theme) {
		case 'light':
			return 'Light';
		case 'dark':
			return 'Dark';
		default:
			return 'System';
	}
}

export type { Theme };
export { themes, defaultTheme, isTheme, getThemeName };
