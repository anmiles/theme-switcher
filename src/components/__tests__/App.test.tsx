import { render, screen, fireEvent, act } from '@testing-library/react';
import type { UserProvider } from '../../providers/userProvider';
import type { SystemProvider } from '../../providers/systemProvider';
import type { AppProps } from '../App';
import App from '../App';
import { EventEmitter } from '../../lib/eventEmitter';
import type { Theme } from '../../lib/theme';
import '@testing-library/jest-dom';

jest.mock('../icons/Checked.tsx', () => function Checked() {
	return '{checked}';
});

jest.mock('../icons/Dark.tsx', () => function Dark() {
	return '{dark}';
});

jest.mock('../icons/Light.tsx', () => function Light() {
	return '{light}';
});

jest.mock('../icons/System.tsx', () => function System() {
	return '{system}';
});

let userTheme: Theme | undefined;
let systemTheme: Theme;

const userEventEmitter   = new EventEmitter<{ change : [Theme | undefined] }>();
const systemEventEmitter = new EventEmitter<{ change : [Theme | undefined] }>();

const userProvider: Partial<InstanceType<typeof UserProvider>> = {
	get : jest.fn().mockImplementation(() => userTheme),

	set : jest.fn().mockImplementation((theme: Theme | undefined) => {
		userEventEmitter['emit']('change', theme);
	}),

	on  : jest.fn().mockImplementation(userEventEmitter.on.bind(userEventEmitter)),
	off : jest.fn().mockImplementation(userEventEmitter.off.bind(userEventEmitter)),
};

const systemProvider: Partial<InstanceType<typeof SystemProvider>> = {
	get   : jest.fn().mockImplementation(() => systemTheme),
	watch : jest.fn(),
	on    : jest.fn().mockImplementation(systemEventEmitter.on.bind(systemEventEmitter)),
	off   : jest.fn().mockImplementation(systemEventEmitter.on.bind(systemEventEmitter)),
};

jest.mock<{ UserProvider : typeof UserProvider }>('../../providers/userProvider', () => ({
	UserProvider : jest.fn().mockImplementation(() => userProvider),
}));

jest.mock<{ SystemProvider : typeof SystemProvider }>('../../providers/systemProvider', () => ({
	SystemProvider : jest.fn().mockImplementation(() => systemProvider),
}));

function renderApp({ float }: AppProps) {
	render(
		<App float={ float } />,
	);
}

afterEach(() => {
	document.body.removeAttribute('data-theme');
});

describe('src/App', () => {
	describe('theme', () => {
		function clickDropdownItem(theme: Theme | 'system'): void {
			renderApp({});

			const themeSwitcher = screen.getByTestId('theme-switcher');

			act(() => {
				fireEvent.click(themeSwitcher);
			});

			const dropdownItemTestId = `theme-item-${theme}`;
			const dropdownItem       = screen.queryByTestId(dropdownItemTestId);

			if (!dropdownItem) {
				throw new Error(`Dropdown does not have an item with test id ${dropdownItemTestId}`);
			}

			act(() => {
				fireEvent.click(dropdownItem);
			});
		}

		it('should be forced to user theme (light)', () => {
			systemTheme = 'dark';
			userTheme   = 'light';

			renderApp({});

			expect(document.body.getAttribute('data-theme')).toBe('light');
		});

		it('should be forced to user theme (dark)', () => {
			systemTheme = 'light';
			userTheme   = 'dark';

			renderApp({});

			expect(document.body.getAttribute('data-theme')).toBe('dark');
		});

		it('should fallback to system theme(light) if user theme is not set', () => {
			userTheme   = undefined;
			systemTheme = 'light';

			renderApp({});

			expect(document.body.getAttribute('data-theme')).toBe('light');
		});

		it('should fallback to system theme(dark) if user theme is not set', () => {
			userTheme   = undefined;
			systemTheme = 'dark';

			renderApp({});

			expect(document.body.getAttribute('data-theme')).toBe('dark');
		});

		it('should be changed on change user provider', () => {
			userTheme = 'light';

			renderApp({});

			expect(document.body.getAttribute('data-theme')).toBe('light');

			act(() => {
				userEventEmitter['emit']('change', 'dark');
			});

			expect(document.body.getAttribute('data-theme')).toBe('dark');
		});

		it('should be changed on change system provider if user theme is not defined', () => {
			userTheme   = undefined;
			systemTheme = 'light';

			renderApp({});

			expect(document.body.getAttribute('data-theme')).toBe('light');

			act(() => {
				systemEventEmitter['emit']('change', 'dark');
			});

			expect(document.body.getAttribute('data-theme')).toBe('dark');
		});

		it('should change to dark if dark item chosen', () => {
			systemTheme = 'light';
			userTheme   = 'light';

			clickDropdownItem('dark');

			expect(document.body.getAttribute('data-theme')).toBe('dark');
		});

		it('should change to light if light item chosen', () => {
			systemTheme = 'dark';
			userTheme   = 'dark';

			clickDropdownItem('light');

			expect(document.body.getAttribute('data-theme')).toBe('light');
		});

		it('should change to system dark if system item chosen', () => {
			systemTheme = 'dark';
			userTheme   = 'light';

			clickDropdownItem('system');

			expect(document.body.getAttribute('data-theme')).toBe('dark');
		});

		it('should change to system light if system item chosen', () => {
			systemTheme = 'light';
			userTheme   = 'dark';

			clickDropdownItem('system');

			expect(document.body.getAttribute('data-theme')).toBe('light');
		});
	});

	describe('icon', () => {
		it('should match light system theme if user theme is not defined', () => {
			userTheme   = undefined;
			systemTheme = 'light';

			renderApp({});

			const themeSwitcher = screen.getByTestId('theme-switcher');
			expect(themeSwitcher.innerHTML).toEqual('{light}');
		});

		it('should match dark system theme if user theme is not defined', () => {
			userTheme   = undefined;
			systemTheme = 'dark';

			renderApp({});

			const themeSwitcher = screen.getByTestId('theme-switcher');
			expect(themeSwitcher.innerHTML).toEqual('{dark}');
		});

		it('should match light user theme despite of system theme', () => {
			userTheme   = 'light';
			systemTheme = 'dark';

			renderApp({});

			const themeSwitcher = screen.getByTestId('theme-switcher');
			expect(themeSwitcher.innerHTML).toEqual('{light}');
		});

		it('should match dark user theme despite of system theme', () => {
			userTheme   = 'dark';
			systemTheme = 'light';

			renderApp({});

			const themeSwitcher = screen.getByTestId('theme-switcher');
			expect(themeSwitcher.innerHTML).toEqual('{dark}');
		});
	});

	describe('switcher', () => {
		describe('html', () => {

			function actOnThemeSwitcher(): HTMLElement {
				renderApp({});

				const themeSwitcher = screen.getByTestId('theme-switcher');

				act(() => {
					fireEvent.click(themeSwitcher);
				});

				return themeSwitcher;
			}

			it('should match snapshot on light system theme', () => {
				userTheme   = undefined;
				systemTheme = 'light';

				expect(actOnThemeSwitcher()).toMatchSnapshot();
			});

			it('should match snapshot on dark system theme', () => {
				userTheme   = undefined;
				systemTheme = 'light';

				expect(actOnThemeSwitcher()).toMatchSnapshot();
			});

			it('should match snapshot on light user theme', () => {
				userTheme   = 'light';
				systemTheme = 'dark';

				expect(actOnThemeSwitcher()).toMatchSnapshot();
			});

			it('should match snapshot on dark user theme', () => {
				userTheme   = 'dark';
				systemTheme = 'light';

				expect(actOnThemeSwitcher()).toMatchSnapshot();
			});
		});

		describe('data-float', () => {
			it('should be set left from float prop', () => {
				renderApp({ float : 'left' });

				const themeSwitcher = screen.getByTestId('theme-switcher');
				expect(themeSwitcher.getAttribute('data-float')).toEqual('left');
			});

			it('should be set right from float prop', () => {
				renderApp({ float : 'right' });

				const themeSwitcher = screen.getByTestId('theme-switcher');
				expect(themeSwitcher.getAttribute('data-float')).toEqual('right');
			});

			it('should be empty if float prop is not set', () => {
				renderApp({});

				const themeSwitcher = screen.getByTestId('theme-switcher');
				expect(themeSwitcher.getAttribute('data-float')).toEqual(null);
			});
		});

		describe('interaction', () => {
			it('should toggle on themeSwitcher click', () => {
				systemTheme = 'light';
				userTheme   = 'dark';

				renderApp({});

				const themeSwitcher = screen.getByTestId('theme-switcher');

				act(() => {
					fireEvent.click(themeSwitcher);
				});

				expect(screen.queryByTestId('theme-selector')).toBeInTheDocument();

				act(() => {
					fireEvent.click(themeSwitcher);
				});

				expect(screen.queryByTestId('theme-selector')).not.toBeInTheDocument();
			});

			it('should hide on dropdown item click', () => {
				systemTheme = 'light';
				userTheme   = 'dark';

				renderApp({});

				const themeSwitcher = screen.getByTestId('theme-switcher');

				act(() => {
					fireEvent.click(themeSwitcher);
				});

				const dropdownItem = themeSwitcher.querySelector('li');

				if (!dropdownItem) {
					throw new Error('Dropdown does not have items');
				}

				act(() => {
					fireEvent.click(dropdownItem);
				});

				expect(screen.queryByTestId('theme-selector')).not.toBeInTheDocument();
			});
		});
	});
});
