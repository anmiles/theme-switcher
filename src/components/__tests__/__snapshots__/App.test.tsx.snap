// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/App switcher html should match snapshot on dark system theme 1`] = `
<div
  class="themeSwitcher"
  data-testid="theme-switcher"
>
  {light}
  <ul
    data-testid="theme-selector"
  >
    <li
      data-testid="theme-item-light"
    >
      {light}
      <span>
        Light
      </span>
    </li>
    <li
      data-testid="theme-item-dark"
    >
      {dark}
      <span>
        Dark
      </span>
    </li>
    <li
      data-testid="theme-item-system"
    >
      {system}
      <span>
        System
      </span>
      {checked}
    </li>
  </ul>
</div>
`;

exports[`src/App switcher html should match snapshot on dark user theme 1`] = `
<div
  class="themeSwitcher"
  data-testid="theme-switcher"
>
  {dark}
  <ul
    data-testid="theme-selector"
  >
    <li
      data-testid="theme-item-light"
    >
      {light}
      <span>
        Light
      </span>
    </li>
    <li
      data-testid="theme-item-dark"
    >
      {dark}
      <span>
        Dark
      </span>
      {checked}
    </li>
    <li
      data-testid="theme-item-system"
    >
      {system}
      <span>
        System
      </span>
    </li>
  </ul>
</div>
`;

exports[`src/App switcher html should match snapshot on light system theme 1`] = `
<div
  class="themeSwitcher"
  data-testid="theme-switcher"
>
  {light}
  <ul
    data-testid="theme-selector"
  >
    <li
      data-testid="theme-item-light"
    >
      {light}
      <span>
        Light
      </span>
    </li>
    <li
      data-testid="theme-item-dark"
    >
      {dark}
      <span>
        Dark
      </span>
    </li>
    <li
      data-testid="theme-item-system"
    >
      {system}
      <span>
        System
      </span>
      {checked}
    </li>
  </ul>
</div>
`;

exports[`src/App switcher html should match snapshot on light user theme 1`] = `
<div
  class="themeSwitcher"
  data-testid="theme-switcher"
>
  {light}
  <ul
    data-testid="theme-selector"
  >
    <li
      data-testid="theme-item-light"
    >
      {light}
      <span>
        Light
      </span>
      {checked}
    </li>
    <li
      data-testid="theme-item-dark"
    >
      {dark}
      <span>
        Dark
      </span>
    </li>
    <li
      data-testid="theme-item-system"
    >
      {system}
      <span>
        System
      </span>
    </li>
  </ul>
</div>
`;
